content/technotes/2018-12-13-pdftools-20.md

```diff
@@ -0,0 +1,91 @@
+
+## Gimme the text
+
+We use an [example pdf file](https://github.com/ropensci/tabulizer/raw/master/inst/examples/data.pdf) from the ropensci tabulizer package. This file contains a few standard datasets which have been written to a table. First let's try the `pdf_text()` function, which returns a character vector of length equal to the number of pages in the file.
+
+
+```r
+library(pdftools)
+pdf_file <- "https://github.com/ropensci/tabulizer/raw/master/inst/examples/data.pdf"
+txt <- pdf_text(pdf_file)
+cat(text[1])
+```
+```
+                    mpg  cyl    disp  hp drat    wt  qsec vs am gear carb
+Mazda RX4           21.0   6   160.0 110 3.90 2.620 16.46  0  1    4    4
+Mazda RX4 Wag       21.0   6   160.0 110 3.90 2.875 17.02  0  1    4    4
+Datsun 710          22.8   4   108.0  93 3.85 2.320 18.61  1  1    4    1
+Hornet 4 Drive      21.4   6   258.0 110 3.08 3.215 19.44  1  0    3    1
+Hornet Sportabout   18.7   8   360.0 175 3.15 3.440 17.02  0  0    3    2
+Valiant             18.1   6   225.0 105 2.76 3.460 20.22  1  0    3    1
+Duster 360          14.3   8   360.0 245 3.21 3.570 15.84  0  0    3    4
+Merc 240D           24.4   4   146.7  62 3.69 3.190 20.00  1  0    4    2
+...
+```
+
+Hence `pdf_text()` converts all text on a page to a large string, which works pretty well. However if you would want to parse this text into a data frame (using e.g. `read.table`) you run into a problem: the first column contains spaces within the values. Therefore we can't use the whitespace as the column delimiter as is the default for `read.table`.
+
+So if we want to write pdf table extractor, we _have_ to infer the column from the _physical position_ of the textbox, rather than rely on delimiting characters. The new `pdf_data()` gives us exactly this. It returns a data frame with all textboxes in a page, including their width, height, and (x,y) position:
+
+
+```r
+pdf_data(pdf_file)[[1]]
+```
+```
+# A tibble: 430 x 6
+   width height     x     y space text  
+   <int>  <int> <int> <int> <lgl> <chr> 
+ 1    29      8   154   139 TRUE  Mazda 
+ 2    19      8   187   139 FALSE RX4   
+ 3    29      8   154   151 TRUE  Mazda 
+ 4    19      8   187   151 TRUE  RX4   
+ 5    19      8   210   151 FALSE Wag   
+ 6    31      8   154   163 TRUE  Datsun
+ 7    14      8   189   163 FALSE 710   
+ 8    30      8   154   176 TRUE  Hornet
+ 9     4      8   188   176 TRUE  4     
+10    23      8   196   176 FALSE Drive 
+```
+
+Converting this pdf data into the original data frame is left as an exercise for the reader :)
```

Comment:
!!! It sounds like a cool puzzle actually!

Reply:
How would this compare to `tabulizer`? You should at least mention `tabulizer` here.